File: real-time-server/src/services/DriverService.js
const { v4: uuidv4 } = require('uuid');

class DriverService {
  constructor(redisClient) {
    this.redis = redisClient;
    this.drivers = new Map();
    this.driverSockets = new Map();
    // Rate limiting: track last location update per driver
    this._lastLocationEmit = new Map();
    this.LOCATION_RATE_LIMIT_MS = 3000; // Min 3s between location broadcasts
  }

  async registerDriver(driverId, socketId, driverData) {
    const driver = {
      id: driverId,
      socketId,
      name: driverData.name || 'Driver',
      phone: driverData.phone || '',
      vehicle: driverData.vehicle || 'Car',
      lat: driverData.lat || 0,
      lng: driverData.lng || 0,
      status: 'available',
      currentOrderId: null,
      lastLocationUpdate: Date.now(),
      rating: driverData.rating || 5.0,
      totalDeliveries: driverData.totalDeliveries || 0
    };

    this.drivers.set(driverId, driver);
    this.driverSockets.set(socketId, driverId);

    if (this.redis && this.redis.isOpen) {
      try {
        await this.redis.hSet(`driver:${driverId}`, {
          ...driver,
          lastLocationUpdate: driver.lastLocationUpdate.toString()
        });
        await this.redis.sAdd('drivers:online', driverId);
        // Store in Redis GEO index for spatial queries
        if (driver.lat && driver.lng) {
          await this.redis.geoAdd('drivers:locations', {
            longitude: driver.lng,
            latitude: driver.lat,
            member: driverId
          });
        }
      } catch (err) {
        console.error(`Redis error registering driver ${driverId}:`, err.message);
      }
    }

    console.log(`Driver ${driverId} registered and online`);
    return driver;
  }

  async updateLocation(driverId, lat, lng) {
    const driver = this.drivers.get(driverId);
    if (driver) {
      driver.lat = lat;
      driver.lng = lng;
      driver.lastLocationUpdate = Date.now();
      this.drivers.set(driverId, driver);

      if (this.redis && this.redis.isOpen) {
        try {
          await this.redis.hSet(`driver:${driverId}`, {
            lat: lat.toString(),
            lng: lng.toString(),
            lastLocationUpdate: driver.lastLocationUpdate.toString()
          });
          // Update Redis GEO index
          await this.redis.geoAdd('drivers:locations', {
            longitude: lng,
            latitude: lat,
            member: driverId
          });
        } catch (err) {
          console.error(`Redis error updating driver ${driverId} location:`, err.message);
        }
      }
    }
    return driver;
  }

  /**
   * Rate-limits location broadcasts to customers.
   * Returns true if the broadcast should proceed.
   */
  shouldBroadcastLocation(driverId) {
    const now = Date.now();
    const last = this._lastLocationEmit.get(driverId) || 0;
    if (now - last < this.LOCATION_RATE_LIMIT_MS) {
      return false;
    }
    this._lastLocationEmit.set(driverId, now);
    return true;
  }

  async setDriverStatus(driverId, status, orderId = null) {
    const driver = this.drivers.get(driverId);
    if (driver) {
      driver.status = status;
      driver.currentOrderId = orderId;
      this.drivers.set(driverId, driver);

      if (this.redis && this.redis.isOpen) {
        try {
          await this.redis.hSet(`driver:${driverId}`, {
            status,
            currentOrderId: orderId || ''
          });
        } catch (err) {
          console.error(`Redis error setting driver ${driverId} status:`, err.message);
        }
      }
    }
    return driver;
  }

  async removeDriver(socketId) {
    const driverId = this.driverSockets.get(socketId);
    if (driverId) {
      this.drivers.delete(driverId);
      this.driverSockets.delete(socketId);
      this._lastLocationEmit.delete(driverId);

      if (this.redis && this.redis.isOpen) {
        try {
          await this.redis.sRem('drivers:online', driverId);
          await this.redis.del(`driver:${driverId}`);
          await this.redis.zRem('drivers:locations', driverId);
        } catch (err) {
          console.error(`Redis error removing driver ${driverId}:`, err.message);
        }
      }

      console.log(`Driver ${driverId} went offline`);
    }
    return driverId;
  }

  async getOnlineDrivers() {
    return Array.from(this.drivers.values()).filter(d => d.status === 'available');
  }

  /**
   * Find nearest available drivers using Redis GEO (O(log(N)) vs O(N) full scan).
   * Falls back to in-memory Haversine if Redis is unavailable.
   */
  async findNearestAvailableDrivers(lat, lng, excludeDriverIds = [], limit = 5) {
    // Try Redis GEO first (much faster for large driver counts)
    if (this.redis && this.redis.isOpen) {
      try {
        const nearbyMembers = await this.redis.geoSearchWith(
          'drivers:locations',
          { longitude: lng, latitude: lat },
          { radius: 50, unit: 'km' },
          ['WITHDIST', 'WITHCOORD'],
          { SORT: 'ASC', COUNT: limit + excludeDriverIds.length }
        );

        const results = [];
        for (const member of nearbyMembers) {
          const driverId = member.member;
          if (excludeDriverIds.includes(driverId)) continue;

          const driver = this.drivers.get(driverId);
          if (driver && driver.status === 'available') {
            results.push({
              ...driver,
              distance: parseFloat(member.distance)
            });
            if (results.length >= limit) break;
          }
        }

        if (results.length > 0) return results;
      } catch (err) {
        console.error('Redis GEO search error, falling back to in-memory:', err.message);
      }
    }

    // Fallback: in-memory Haversine
    const availableDrivers = Array.from(this.drivers.values())
      .filter(d => d.status === 'available' && !excludeDriverIds.includes(d.id));

    const driversWithDistance = availableDrivers.map(driver => ({
      ...driver,
      distance: this.calculateDistance(lat, lng, driver.lat, driver.lng)
    }));

    driversWithDistance.sort((a, b) => a.distance - b.distance);

    return driversWithDistance.slice(0, limit);
  }

  calculateDistance(lat1, lng1, lat2, lng2) {
    const R = 6371;
    const dLat = this.toRad(lat2 - lat1);
    const dLng = this.toRad(lng2 - lng1);
    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(this.toRad(lat1)) * Math.cos(this.toRad(lat2)) *
      Math.sin(dLng / 2) * Math.sin(dLng / 2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
  }

  toRad(deg) {
    return deg * (Math.PI / 180);
  }

  getDriverBySocketId(socketId) {
    const driverId = this.driverSockets.get(socketId);
    return driverId ? this.drivers.get(driverId) : null;
  }

  getDriverById(driverId) {
    return this.drivers.get(driverId);
  }

  getDriverSocketId(driverId) {
    const driver = this.drivers.get(driverId);
    return driver ? driver.socketId : null;
  }
}

module.exports = DriverService;

File: backend/realtime-service/internal/driverservice.go
package internal

import (
	"context"
	"log"
	"sort"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
	"zimfeast/shared/geo"
	"zimfeast/shared/redispub"
)

type DriverInfo struct {
	ID            string                 `json:"id"`
	Name          string                 `json:"name"`
	Phone         string                 `json:"phone"`
	Vehicle       map[string]interface{} `json:"vehicle"`
	Lat           float64                `json:"lat"`
	Lng           float64                `json:"lng"`
	Status        string                 `json:"status"`
	SocketID      string                 `json:"socket_id"`
	OrderID       string                 `json:"current_order_id"`
	LastBroadcast time.Time              `json:"-"`
}

type DriverService struct {
	mu      sync.RWMutex
	drivers map[string]*DriverInfo
	sockets map[string]string
	pub     *redispub.Publisher
}

func NewDriverService(pub *redispub.Publisher) *DriverService {
	return &DriverService{
		drivers: make(map[string]*DriverInfo),
		sockets: make(map[string]string),
		pub:     pub,
	}
}

func (ds *DriverService) RegisterDriver(driverID, socketID, name, phone string, lat, lng float64, vehicle map[string]interface{}) {
	ds.mu.Lock()
	defer ds.mu.Unlock()

	ds.drivers[driverID] = &DriverInfo{
		ID:       driverID,
		Name:     name,
		Phone:    phone,
		Vehicle:  vehicle,
		Lat:      lat,
		Lng:      lng,
		Status:   "available",
		SocketID: socketID,
	}
	ds.sockets[socketID] = driverID

	ctx := context.Background()
	ds.pub.Client().GeoAdd(ctx, "drivers:locations", &redis.GeoLocation{
		Name:      driverID,
		Latitude:  lat,
		Longitude: lng,
	})
	ds.pub.Client().SAdd(ctx, "drivers:online", driverID)
	ds.pub.Client().HSet(ctx, "driver:"+driverID, map[string]interface{}{
		"name": name, "phone": phone, "status": "available",
		"lat": lat, "lng": lng, "socketId": socketID,
	})

	log.Printf("[driver] registered: %s (%s)", driverID, name)
}

func (ds *DriverService) UpdateLocation(driverID string, lat, lng float64) {
	ds.mu.Lock()
	defer ds.mu.Unlock()

	d, ok := ds.drivers[driverID]
	if !ok {
		return
	}
	d.Lat = lat
	d.Lng = lng

	ctx := context.Background()
	ds.pub.Client().GeoAdd(ctx, "drivers:locations", &redis.GeoLocation{
		Name:      driverID,
		Latitude:  lat,
		Longitude: lng,
	})
	ds.pub.Client().HSet(ctx, "driver:"+driverID, "lat", lat, "lng", lng)
}

func (ds *DriverService) ShouldBroadcastLocation(driverID string) bool {
	ds.mu.RLock()
	defer ds.mu.RUnlock()
	d, ok := ds.drivers[driverID]
	if !ok {
		return false
	}
	return time.Since(d.LastBroadcast) >= 3*time.Second
}

func (ds *DriverService) MarkBroadcast(driverID string) {
	ds.mu.Lock()
	defer ds.mu.Unlock()
	if d, ok := ds.drivers[driverID]; ok {
		d.LastBroadcast = time.Now()
	}
}

func (ds *DriverService) SetDriverStatus(driverID, status string) {
	ds.mu.Lock()
	defer ds.mu.Unlock()
	d, ok := ds.drivers[driverID]
	if !ok {
		return
	}
	d.Status = status
	ctx := context.Background()
	ds.pub.Client().HSet(ctx, "driver:"+driverID, "status", status)
	if status == "available" {
		ds.pub.Client().SAdd(ctx, "drivers:online", driverID)
	} else if status == "offline" {
		ds.pub.Client().SRem(ctx, "drivers:online", driverID)
	}
}

func (ds *DriverService) SetDriverOrder(driverID, orderID string) {
	ds.mu.Lock()
	defer ds.mu.Unlock()
	if d, ok := ds.drivers[driverID]; ok {
		d.OrderID = orderID
		d.Status = "busy"
	}
}

func (ds *DriverService) ClearDriverOrder(driverID string) {
	ds.mu.Lock()
	defer ds.mu.Unlock()
	if d, ok := ds.drivers[driverID]; ok {
		d.OrderID = ""
		d.Status = "available"
	}
}

func (ds *DriverService) GetDriver(driverID string) *DriverInfo {
	ds.mu.RLock()
	defer ds.mu.RUnlock()
	if d, ok := ds.drivers[driverID]; ok {
		cp := *d
		return &cp
	}
	return nil
}

func (ds *DriverService) GetDriverBySocket(socketID string) *DriverInfo {
	ds.mu.RLock()
	defer ds.mu.RUnlock()
	driverID, ok := ds.sockets[socketID]
	if !ok {
		return nil
	}
	if d, ok := ds.drivers[driverID]; ok {
		cp := *d
		return &cp
	}
	return nil
}

func (ds *DriverService) RemoveDriver(socketID string) {
	ds.mu.Lock()
	defer ds.mu.Unlock()
	driverID, ok := ds.sockets[socketID]
	if !ok {
		return
	}
	delete(ds.sockets, socketID)
	delete(ds.drivers, driverID)
	ctx := context.Background()
	ds.pub.Client().SRem(ctx, "drivers:online", driverID)
	ds.pub.Client().ZRem(ctx, "drivers:locations", driverID)
	ds.pub.Client().Del(ctx, "driver:"+driverID)
	log.Printf("[driver] removed: %s", driverID)
}

func (ds *DriverService) GetOnlineDrivers() []*DriverInfo {
	ds.mu.RLock()
	defer ds.mu.RUnlock()
	var result []*DriverInfo
	for _, d := range ds.drivers {
		if d.Status == "available" {
			cp := *d
			result = append(result, &cp)
		}
	}
	return result
}

func (ds *DriverService) FindNearestAvailableDrivers(lat, lng float64, limit int) []*DriverInfo {
	ds.mu.RLock()
	defer ds.mu.RUnlock()

	type dd struct {
		driver *DriverInfo
		dist   float64
	}
	var candidates []dd
	for _, d := range ds.drivers {
		if d.Status != "available" {
			continue
		}
		dist := geo.HaversineDistance(lat, lng, d.Lat, d.Lng)
		candidates = append(candidates, dd{driver: d, dist: dist})
	}
	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i].dist < candidates[j].dist
	})
	var result []*DriverInfo
	for i, c := range candidates {
		if i >= limit {
			break
		}
		cp := *c.driver
		result = append(result, &cp)
	}
	return result
}
